EVA-1586 First version of the pipeline

mistyskye opened the pull request.

mistyskye committed:
Initial commit

mistyskye committed:
Updated SNP filter + header generation

mistyskye committed:
Added reverse strand allele correction & command line execution

Reverse strand allele correction:
If variants are mapped onto the reverse strand of the new assembly, the REF alleles are now ajusted accordingly, (example: the REF allele A becomes T) so that the VCF file always describes variants for the forward strand.
Command line execution:
The pipeline can now be executed via command line, see README for instructions.
Additional features:
Added calculation of the % of remapped variants (output in console).

mistyskye committed:
Fixed typos and bugs to do with the index

mistyskye committed:
Fixed reverse strand positions + sorted bam

In reverse_strand.py, if a variant is mapped to the reverse strand, the position is now +51 and not +50.
In remapping_commands.sh, the sorted bam is now used to generate the vcf file, to go with the sorted bai file.

jmmut reviewed (approved):
I don't have experience using these tools but looking at the intermediate files everything looks reasonable, but I would like to know why remapping the reference reads and adding the alt allele at the end is better than remapping the alternate reads.

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
 # variant-remapping
-The pipeline for remapping VCF variants between two arbitrary FASTA assemblies.
+## Description
+The pipeline for remapping VCF variants between two arbitrary FASTA assemblies. No chain file required.
```
```suggestion
Pipeline for remapping VCF variants between two arbitrary assemblies in FASTA format. No chain file is required.
```

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
 # variant-remapping
-The pipeline for remapping VCF variants between two arbitrary FASTA assemblies.
+## Description
```
For consistency with other README files.
```suggestion
```

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
+- bcftools
+
+## Input
+- Old genome assembly file (fasta format): the genome you have variants for.
```
```suggestion
- Old genome assembly file (FASTA format): the genome you have variants for.
```

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
+
+## Input
+- Old genome assembly file (fasta format): the genome you have variants for.
+- New genome assembly file (fasta format): the genome you want to remap the variants to.
```
```suggestion
- New genome assembly file (FASTA format): the genome you want to remap the variants to.
```

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
+## Input
+- Old genome assembly file (fasta format): the genome you have variants for.
+- New genome assembly file (fasta format): the genome you want to remap the variants to.
+- Variants file (vcf format): contains the list of variants you want to remap.
```
```suggestion
- Variants file (VCF format): contains the list of variants you want to remap.
```

cyenyxe commented on README.md:
````diff
@@ -1,2 +1,89 @@
+sys	0m0.171s
+```
+
+## Task list
````
Shall we create tickets in the backlog? In terms of tracking and planning, it is preferable over a list in a plain text file :)

mistyskye replied:
Yes that would be best :) Should I make them?

tskir replied:
Since at this point we don't have a definitive long-term roadmap for the pipeline, I think it's too soon to start creating the tickets. The only thing that we'll surely have to implement (during @mistyskye's internship or after it) is support for indels. Other points are more planning points than tickets; so I suggest just moving them to the shared Google document so far

tskir replied:
@mistyskye Could you please remove the task list from the `README` and move it to the Google doc (if not already there)

cyenyxe commented on README.md:
```diff
@@ -1,2 +1,89 @@
+Currently, it only supports SNPs, and not indels.
+
+Prerequisites:
+- reverse_strand.py, included (reverse strand allele correction, uses bamnostic)
```
Does this mean the alleles are converted to the forward strand? Could it be written down that way, for the not-so-biologically savvy (most people in the team 😂)?

mistyskye replied:
Yes! I've added an explanation in the README file.

cyenyxe commented on remapping_commands.sh:
```diff
@@ -0,0 +1,189 @@
+bedtools flank -i variants.bed -g "$oldgenome".chrom.sizes -b 50 > flanking.bed
+rm variants.bed "$oldgenome".chrom.sizes
+
+#fix rows: put both flanking sequences on 1 row = 1 read and add 1 to start position so the read is 100 bases:
```
Could you please also document briefly what it looked like before, to understand better the reason for these transformations?

mistyskye replied:
Of course, I have now added an explanation directly into the remapping_commands.sh file.

cyenyxe commented on remapping_commands.sh:
```diff
@@ -0,0 +1,189 @@
+bedtools getfasta -fi $oldgenome -bed flanking.corrected.bed -fo variants_reads.fa
+
+#replace the colon separators with "|"
+sed 's/:/|/' variants_reads.fa > variants_reads2.fa
```
`sed` support in-place replacements with the `-i` argument.

mistyskye replied:
Thank you, I changed the command.

cyenyxe commented on remapping_commands.sh:
```diff
@@ -0,0 +1,189 @@
+#convert vcf to bed:
+vcf2bed < snps_only.vcf > variants.bed
+#the actual position of the variant is the second coordinate
+#rm snps_only.vcf
```
Please delete commands that are not planned to be used. There are a few more in step 3.

mistyskye replied:
I commented the rm commands for testing purposes, but I forgot to uncomment them! I'll fix them.

cyenyxe commented on reverse_strand.py:
```diff
@@ -0,0 +1,35 @@
+	info=name.split("|")
+	#mapped onto reverse strand:
+	if read.flag & 16: #decoding bitwise flag with &; may need to be changed back to ==
+		if info[2]=='A':
```
My suggestion is to assign the reverse nucleotide in the if-else, and take the long `write` call out of them to avoid repetition.

mistyskye replied:
Good suggestion, added.

tskir replied:
@mistyskye Although this discussion is marked as resolved, I can still see two `outfile.write()` calls in this PR. Can you please check if you have pushed all of the latest commits?

mistyskye replied:
Oh I didn't change the else part, this has been fixed.

cyenyxe commented on write_header.py:
```diff
@@ -0,0 +1,40 @@
+import argparse
+from argparse import RawTextHelpFormatter
+
+description="Adds the first part of a VCF header up until the end of the ##INFOs, \nthen adds the newly generated ##configs and ##reference, and finally add the rest of the header."
```
A header can be sorted any way, which means there could be useful information after an INFO entry.

mistyskye replied:
I wasn't sure about the order of VCF headers, so if they're un-ordered does this mean I can just delete lines starting with ##contig from the old header, regenerate them using the new contig names, and paste them after the old header? In that case, then I agree that this Python script is unnecessary.

tskir replied:
Meta-information (`##`) lines can in theory go in any order, except for the very first `##fileformat` line. In the VCF specification, sections describing different meta-information lines are arranged in some order, but it never says that lines themselves must be arranged: https://samtools.github.io/hts-specs/VCFv4.2.pdf

I suggest you use `bcftools reheader` for replacing contig names

mistyskye replied:
I've changed how I generate the header, it now just simply adds everything that isn't a contig to my newly generated contigs list to make the new header. If there is still a problem with this, let me know and I will use `bcftools reheader`.

cyenyxe commented on reverse_strand.py:
```diff
@@ -0,0 +1,35 @@
+bamfile=bs.AlignmentFile(args.bam, 'rb')
+outfile=open(args.outfile, 'w')
+
+#reverses the allele for variants that got mapped onto the reverse strand of the new genome, and prints everything into correct columns
```
This loop seems to complement the bases, but does it also reverse the sequence? Or is it not needed?

mistyskye replied:
By sequence do you mean the (new) reference genome? Or the read that got mapped?

tskir replied:
I guess @cyenyxe meant the read. Because we're only interested in SNPs for now, we only need to flip the strand for one specific nucleotide, so actually reverse-complementing the read is unnecessary for now

cyenyxe commented on remapping_commands.sh:
```diff
@@ -0,0 +1,189 @@
+
+#find the end of the INFO tags and print the contigs (using custom python script)
+chmod a+x write_header.py
+./write_header.py -i temp_header.txt -c contigs.txt -o final_header.txt
```
If `vcf_header.txt` was created using only lines starting with `##`, then this Python script might not even be necessary because getting the final header would be just a matter of concatenating a bunch of other files.

mistyskye replied:
See [other comment.](https://github.com/EBIvariation/variant-remapping/pull/1/files#r308158748)

mistyskye committed:
Fixed README.md (formatting)

Co-Authored-By: Cristina Yenyxe Gonzalez Garcia <cyenyxe@ebi.ac.uk>

mistyskye committed:
Fixed bugs, changed parts based on suggestions

-Fixed problem with grep command not being inclusive enough (remapping_commands.sh)
-Added explanations for unclear sections (README.md and remapping_commands.sh)
-Changed sed command (remapping_commands.sh)
-Deleted unused commands, uncommented rm cleanup commands (remapping_commands.sh)
-Cleaned up code in reverse_strand.py
-Changed header generation: write_header.py is no longer needed

mistyskye committed:
Resolved merge conflicts

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+# -copy vcf header to the new vcf file, but regenerate contigs from the list of contigs in new assembly, and change reference tag
+# -take into account reverse strand
+# -create a command line version where you input file names
+# -% of remapped variants calculation
```
Given that we have a separate Markdown file with documentation, which already includes most if not all of these comments, they can be safely removed from here

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+done
+
+
+#1) Flanking sequence generation
```
Since you have quite prominent `echo` statements which already define the steps, comments such as this one can be removed

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+echo "-----------------------1) Flanking sequence generation-----------------------"
+#generate chrom sizes:
+samtools faidx $oldgenome 
+cut -f1,2 "$oldgenome".fai > "$oldgenome".chrom.sizes
```
In this example, `$oldgenome` is double-quoted in the second line but not in the first one. For simplicity and reliability, I suggest double-quoting all variable expansions in all contexts throughout the script

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+bcftools filter -i 'TYPE="snp"' $vcffile > snps_only.vcf
+
+#store header
+awk '$1 ~ /^#/ {print $0}' snps_only.vcf > vcf_header.txt
```
FYI this can be achieved more simply by `grep '^#'` :) However, the problem with the awk/grep approach is that they will scan the entire file, which can be millions of line long. It's better to use `bcftools view --header-only` to extract the header

tskir replied:
(It will only scan up until the `#CHROM POS ...` line and then quit)

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+#convert vcf to bed:
+vcf2bed < snps_only.vcf > variants.bed
+#the actual position of the variant is the second coordinate
+rm snps_only.vcf
```
Given that we create *lots* of temporary files, I think it's better to store them in a temporary directory instead of just putting them in the current working directory. You can use `mktemp -d -t` to create such a directory

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+rm snps_only.vcf
+
+#generate the flanking sequence intervals
+bedtools flank -i variants.bed -g "$oldgenome".chrom.sizes -b 50 > flanking.bed
```
Theoretically, even though unlikely, it is possible for a variant to be so close to the end of the chromosome that it's impossible to get a 50 bp flanking sequence for it. Could you please investigate what would happen in such a case when trying to run this command?

_This suggestion is low-priority_

mistyskye replied:
Bedtools flank requires a file with the length of each chromosome/contig (.chrom.sizes), this is used exactly for this purpose: 
[From the bedtools website:](https://bedtools.readthedocs.io/en/latest/content/tools/flank.html)
> Note that flank will restrict the created flanking intervals to the size of the chromosome (i.e. no start < 0 and no end > chromosome size).

So the read would be shorter, but no errors would occur.

tskir replied:
Right, but then this would cause problems later in the pipeline because we expect the SNP to always be at 51st position of the read; if the read is shorter the offset will be different. We should probably add a step where all reads shorter than the expected length (101) are filtered out prior to alignment

tskir replied:
(This remains low-priority)

mistyskye replied:
I added a filter: all reads not equal to 101 bases are now filtered out.

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+#fix rows: put both flanking sequences on 1 row so each read is 101 bases: [50 bases]-[variant]-[50 bases]
+awk 'NR == 1 {prev = $2} NR>1 && NR%2==0 {print $1"\t"prev"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7"\t"$8"\t"$9} NR%2 == 1 { prev = $2}' flanking.bed > flanking.corrected.bed
+rm flanking.bed
```
You can use `bedtools slop` instead of `bedtools flank` to generate a two-way flanking sequence in one go

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+bedtools getfasta -fi $oldgenome -bed flanking.corrected.bed -fo variants_reads.fa
+
+#replace the colon separators with "|"
+sed -i 's/:/|/' variants_reads.fa
```
I'm a bit lost here, why is this step necessary? Is there some specific format of the sequence names?

mistyskye replied:
I'm storing this information for later on in the script when I split the name by "|" to extract the relevant information (ALT allele, QUAL, FILT, INFO). I do this because the INFO column can contain ":", which means I wouldn't be able to split by ":", so I chose to use "|". Should I add this as an explanation in the script?

tskir replied:
Got it. Yes, please add an explanation to the code

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+#reformat the fasta ID
+sed 's/\t/|/g; s/ /|/g; s/\(.*\)|/\1 /' temp.txt | tr ' ' '\n' > variant_reads.out.fa
+rm temp.txt
```
Could you please elaborate what's happening here? It would help to provide examples of input/output lines as done earlier in the script

tskir reviewed (changes requested):
The review is not yet complete; only covers documentation and Part 1 of the `remapping_commands.sh`

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+
+
+#2) Mapping with bowtie2
```
I feel that 6 empty lines are a bit excessive :) A double empty line would suffice to separate the larger code blocks in my opinion

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+echo "#### Building index: may take a while ####"
+bowtie2-build $newgenome index
+echo "#### Mapping results: ####"
+bowtie2 -f -x index variant_reads.out.fa | samtools view -bS - > reads_aligned.bam
```
1) Could you please provide an explanation of the `bowtie21 flags used?
2) Although we have observed that Bowtie only stores one alignment per “read”, it's better to specify this behaviour explicitly (using flags)

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+rm reads_aligned.bam reads_aligned.sorted.bam reads_aligned.sorted.bam.bai
+
+#add interval for variant position in bed format (required by getfasta)
+awk '{print $1"\t"$2-1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7}' variants_remapped.vcf > variants_remapped.bed
```
Again this is a bit obscure, could you please provide an input-output example here?

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+#paste all the information into the correct columns
+awk '{print $1"\t"$2"\t"$3}' variants_remapped.vcf > temp_first_3_columns.txt
+awk '{print $4"\t"$5"\t"$6"\t"$7}' variants_remapped.vcf > temp_last_4_columns.txt
```
When you only need to extract some columns, without doing any more complex transformations, it's easier to use `cut -f1-3` or `cut -f4-7`, respectively

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+
+#header:
+#create list of contigs/chromosomes to be added to the header
```
I suggest you look into `bcftools reheader` to update the header information. Among other things, it can automatically update the list of contigs based on fai index. This would also the problem in current implementation where VCF header will only include contigs which have at least one variant on them (in reality it should include **all** contigs present in the assembly)

mistyskye replied:
This has been discussed and we established that we can't use `bcftools reheader` in its current version as it does not have the functionality we need, so I'll continue using the bash commands to recreate the header.

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,181 @@
+
+echo "-----------------------------------Results-----------------------------------"
+echo "Total number of input variants:"
+grep "^[^#]" $vcffile | wc -l
```
At least some of these statistics can be computed by `bcftools stats` instead
https://samtools.github.io/bcftools/bcftools.html#stats

mistyskye replied:
I have added this command to the pipeline, although I have left my results calculations as they are a quick way to get the numbers I need to fill in the spreadsheet.

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,36 @@
+from argparse import RawTextHelpFormatter
+import bamnostic as bs
+
+description="Reverses the allele for variants that got mapped onto the reverse strand of the new genome, and prints everything to a new file\nAlso filters out the non-mapped"
```
Could you please wrap the line length throughout the code to the maximum of 120 columns?

mistyskye replied:
I'm not sure how to set the maximum in Sublime Text 3, however I have set `word_wrap true` which adds auto line wrapping to the files.

tskir replied:
How to set maximum line length in Sublime: https://stackoverflow.com/questions/25900954/80-characters-right-margin-line-in-sublime-text-3 (use 120 instead of 80 as described there)

It looks like the option you currently have enabled wraps the lines to some other limit (or doesn't at all), bercause for example line 157 of the `remapping_commands.sh` is 145 characters long

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,36 @@
+		outfile.write("%s\t%s\t%s\t%s\t%s\t%s\t%s\n" % (read.reference_name,read.pos+51,info[3],rev_nucl,info[4],info[5],info[6]))
+	#only mapped onto direct strand:
+	elif read.flag==0:
+		outfile.write("%s\t%s\t%s\t%s\t%s\t%s\t%s\n" % (read.reference_name,read.pos+51,info[3],info[2],info[4],info[5],info[6]))
```
This can be shortened by using https://biopython.org/wiki/Seq#complement-and-reverse-complement instead of the if-elif loop, and also only having one `outfile.write()` call. E. g.:

```python
nucl = info[2]
if read.flag & 16:
    nucl = Seq(nucl, generic_dna).complement()
outfile.write(...)
```

mistyskye replied:
I agree with the biopython addition. I have added it, along with the single `oufile.write`, although I do need an extra check before it so that I don't write the unmapped (flag=4) variants. (see new version when I push it)

tskir reviewed (changes requested):
The second part of the review

mistyskye committed:
Changed how REF=ALT variants are dealt with, various other fixes

-If REF=ALT after remapping, the old REF allele is used instead of the new REF, and the alleles are then swapped
-Removed tasklist from remapping_commands.sh
-Removed commented titles from remapping_commands.sh
-Added double-quotes on all variable expansions
-Changed header extraction: now uses bcftools view --header-only to avoid scanning the entire file
-Replaced bedtools flank with bedtools slop, removed now obsolete commands
-Added explanations for command flags and commands
-Pipeline now uses a temp dir for all the intermediate files

mistyskye committed:
Added various suggestions, fixed a couple of bugs

-Added explanations for certain commands (remapping_commands.sh)
-Replaced simple awk printing columns with cut (remapping_commands.sh)
-Fixed writing problem in reverse_strand.py
-Shortened code in reverse_strand.py
-reverse_strand.py now uses pysam, and filters out unmapped reads (flag 4)
-Wrapped line length in script files

mistyskye committed:
Added line wrapping + removed task list from README

Manually edited long lines (>120 characters) to get them to wrap onto a new line
Removed the task list from the README, it is now in the Google Doc

mistyskye committed:
Added filter for short reads

Truncated reads are now filtered out: if the variant is too close to the beginning or end of the chromosome¸ the
generated read ends up being shorter (<101 bases). Even though this is a rare occurrence, these reads will now be
removed and will not be remapped.

tskir commented on README.md:
```diff
@@ -1,2 +1,90 @@
+
+**Example command:**
+`time ./remapping_commands.sh -g droso_dm3.fasta -n droso_dm6.fasta -a GCA_000001215.4 -v droso_variants_renamed.vcf \
+	-o test.vcf`
```
It's better to reformat it to make each flag and its parameter start with the new line, like so:

```bash
time ./remapping_commands.sh \
  -g droso_dm3.fasta \
  -n droso_dm6.fasta \
  -a GCA_000001215.4 \
  -v droso_variants_renamed.vcf \
  -o test.vcf
```

tskir replied:
(It makes the argument list more readable)

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,195 @@
+
+#example command:
+#time ./remapping_commands.sh -g droso_dm3.fasta -n GCA_000001215.4_Release_6_plus_ISO1_MT_genomic.fna -a 
+#GCA_000001215.4 -v droso_variants_renamed.vcf -o test.vcf
```
All comments from the beginning of the file up until this line can be removed because they are already covered in the documentation

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,195 @@
+TMPDIR="$(mktemp -d -t --tmpdir=$PWD remapping_tmp_dir_XXX)"
+cd "$TMPDIR"
+echo "-----------------------1) Flanking sequence generation-----------------------"
+#generate chrom sizes:
```
Minor style comment: it is preferable to leave a space between `#` and the actual comment, and starting the comment with the capital letter, e. g.:

```bash
# Generate chrom sizes
```

Could you please fix this throughout the code?

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,195 @@
+#samtools view arguments:
+#-b: output in BAM format
+#-S: used to specifiy input is SAM format, but only in older versions of samtools
+bowtie2 -k 1 -f -x index "$TMPDIR"/variant_reads.out.fa | samtools view -bS - > "$TMPDIR"/reads_aligned.bam
```
It would be helpful for readability to separate different subsections of this section with single empty lines, as you did in the previous section

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,195 @@
+
+#test each variant to see if REF = ALT, if so, replace the current REF with the corresponding old REF (this is to deal 
+#with variants like G > G)
+.././replace_refs.py -i vcf_out_with_header.vcf -r old_ref_alleles_with_header.txt -o pre_final_vcf.vcf
```
`.././` is exactly equivalent to `../`, so this can be shortened

tskir commented on remapping_commands.sh:
```diff
@@ -0,0 +1,195 @@
+percentage=$(bc <<< "scale=3;($remapped/$total)*100")
+echo $percentage"%"
+
+cd ../
```
In the future we might add the possibility for a user to specify the location of the desired temporary directory. In this case, `cd ../` will not necessarily return us to where we were. It's more reliable to use `cd -` (go back to the previous directory)

tskir commented on replace_refs.py:
```diff
@@ -0,0 +1,30 @@
+parser = argparse.ArgumentParser(description = description, formatter_class=RawTextHelpFormatter)
+parser.add_argument("-i", "--vcf", help="VCF file containing remapped variants")
+parser.add_argument("-r", "--oldrefalleles", help="file containing the same header as the VCF file + the old REF \
+        alleles in the same order as the VCF")
```
When splitting string literals, it's not necessary (and actually discouraged) to use the backslash sign. It's better to do it this way:

```python
some_call("some_string_literal_"
          "continuation_of_literal")
```

This will create `some_string_literal_continuation_of_literal`

tskir commented on replace_refs.py:
```diff
@@ -0,0 +1,30 @@
+                                        variant[6],variant[7]))
+                        else:
+                                out_vcf.write(line)
+out_vcf.close()
```
You can add out_vcf declaration to the context as well:

```python
with open(args.vcf, 'r') as vcf, open(args.outfile, 'w') as out_vcf:
    ...
```

In this case, there'll be no need for closing the file explicitly.

tskir commented on replace_refs.py:
```diff
@@ -0,0 +1,30 @@
+out_vcf=open(args.outfile, 'w')
+
+with open(args.vcf, 'r') as vcf:
+        for(i, line) in enumerate(vcf):
```
Our code style (consistent with PEP 8) is to indent using 4 spaces for each level (not 8 spaces as here)

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+#! /home/juliette/miniconda3/bin/python3
```
This shebang is not going to be supported on any other machine than your own one; I suggest using the most universal one, `#! /usr/bin/env python3`

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+#into correct columns
+
+for read in bamfile:
+		name=read.query_name
```
Same comment about indentation; needs to be 4 spaces per level

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+		if read.flag & 16: #decoding bitwise flag with &; may need to be changed back to ==
+			nucl = Seq(nucl, generic_dna).complement()
+		#write it all to the file:
+		if read.flag != 4: #4 means unmapped
```
This needs to use bitwise operators as well, because the read can be unmapped (4) and also have some other flags set

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+for read in bamfile:
+		name=read.query_name
+		info=name.split("|")
+		nucl=info[3]
```
Code style: space around the assignment operator (`nucl = info[3]`)

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+		if read.flag != 4: #4 means unmapped
+			outfile.write("%s\t%s\t%s\t%s\t%s\t%s\t%s\n" % (read.reference_name,read.pos+51,info[4],nucl,info[5],
+				info[6],info[7]))
+			#store old reference allele:
```
Comment style for PEP 8 (here and throughout the code) is to leave space and start with the capital letter, e. g.:
```python
# Store old reference allele
```

tskir reviewed (changes requested):
Final comments, mostly on the code style, after this we can merge this PR as the first version of the pipeline is mostly ready

mistyskye committed:
Fixed minor style issues

- README: changed format of the example command
- remapping_commands.sh: removed comments from the beginning of the file that are now in the README
- Changed "#text" to "# Text" comments in all files, and other additions to do with code style
- Added spacing
- Changed "cd ../" to "cd -"
- Changed incorrect indentations
- Fixed shebang to be universal in reverse_strand.py
- Added bitwise operator for the 4 flag in reverse_strand.py

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+# into correct columns
+
+for read in bamfile:
+	name = read.query_name
```
This file is still using tabs for indentation; it should be 4 spaces instead

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+	if read.flag & 16: # Decoding bitwise flag with &; may need to be changed back to ==
+		nucl = Seq(nucl, generic_dna).complement()
+	# Write it all to the file:
+	if not(read.flag & 4): # 4 means unmapped
```
These can be checked more directly using [`read.is_reverse`](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.is_reverse) and [`read.is_unmapped`](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.is_unmapped) attributes

tskir commented on replace_refs.py:
```diff
@@ -0,0 +1,27 @@
+
+parser = argparse.ArgumentParser(description = description, formatter_class=RawTextHelpFormatter)
+parser.add_argument("-i", "--vcf", help="VCF file containing remapped variants")
+parser.add_argument("-r", "--oldrefalleles", help="file containing the same header as the VCF file + the old" 
```
```suggestion
parser.add_argument("-r", "--oldrefalleles", help="file containing the same header as the VCF file + the old "
```

Space is required, otherwise it will concatenate as "oldREF"

tskir commented on replace_refs.py:
```diff
@@ -0,0 +1,27 @@
+            # Gets corresponding REF allele from the oldrefalleles file
+                out_vcf.write("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s" % (variant[0],variant[1],variant[2],
+                    linecache.getline(args.oldrefalleles, i+1).rstrip(),variant[4],variant[5],
+                    variant[6],variant[7]))
```
Arguments should be separated with spaces:

```python
variant[0], variant[1], variant[2], ...
```

tskir commented on reverse_strand.py:
```diff
@@ -0,0 +1,39 @@
+
+bamfile=pysam.AlignmentFile(args.bam, 'rb')
+outfile=open(args.outfile, 'w')
+old_ref_alleles=open(args.old_ref_alleles, 'w')
```
Space around the assignment operator (here and throughout the code)

tskir commented on README.md:
```diff
@@ -1,2 +1,94 @@
+  -n droso_dm6.fasta \
+  -a GCA_000001215.4 \
+  -v droso_variants_renamed.vcf \
+  -o test.vcf`
```
I don't think this will properly format because single backticks are only meant to be used for inline code formatting. Here it's better to use block code formatting with triple backticks (as shown here https://help.github.com/en/articles/creating-and-highlighting-code-blocks)

tskir reviewed (changes requested):
Some final comments, mostly on code style

mistyskye committed:
Fixed minor spacing issues and other formatting problems

- Fixed tabs as indentation: they are now spaces
- Replaced the flag checkers with read.is_reverse and read.is_unmapped
- Added spacing
- Added code block in README
